mixer/pkg/runtime2/runtime.go

```diff
@@ -0,0 +1,195 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package runtime2
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"istio.io/istio/mixer/pkg/adapter"
+	"istio.io/istio/mixer/pkg/config/store"
+	"istio.io/istio/mixer/pkg/il/compiled"
+	"istio.io/istio/mixer/pkg/pool"
+	"istio.io/istio/mixer/pkg/runtime"
+	"istio.io/istio/mixer/pkg/runtime2/config"
+	"istio.io/istio/mixer/pkg/runtime2/dispatcher"
+	"istio.io/istio/mixer/pkg/runtime2/handler"
+	"istio.io/istio/mixer/pkg/runtime2/routing"
+	"istio.io/istio/mixer/pkg/template"
+	"istio.io/istio/pkg/log"
+)
+
+// Runtime is the main entry point to the Mixer runtime environment. It listens to configuration, instantiates handler
+// instances, creates the dispatch machinery and handles incoming requests.
+type Runtime struct {
+	identityAttribute string
+
+	defaultConfigNamespace string
+
+	ephemeral *config.Ephemeral
+
+	snapshot *config.Snapshot
+
+	handlers *handler.Table
+
+	dispatcher *dispatcher.Dispatcher
+
+	env adapter.Env
+
+	store store.Store
+
+	handlerPool *pool.GoroutinePool
+
+	stateLock            sync.Mutex
+	shutdown             chan struct{}
+	waitQuiesceListening sync.WaitGroup
+}
+
+// New returns a new instance of Runtime.
+func New(
+	s store.Store,
+	templates map[string]*template.Info,
+	adapters map[string]*adapter.Info,
+	identityAttribute string,
+	defaultConfigNamespace string,
+	executorPool *pool.GoroutinePool,
+	handlerPool *pool.GoroutinePool,
+	enableTracing bool) *Runtime {
+
+	runtime := &Runtime{
+		identityAttribute:      identityAttribute,
+		defaultConfigNamespace: defaultConfigNamespace,
+		ephemeral:              config.NewEphemeral(templates, adapters),
+		snapshot:               config.Empty(),
+		handlers:               handler.Empty(),
+		dispatcher:             dispatcher.New(identityAttribute, executorPool, enableTracing),
+		handlerPool:            handlerPool,
+
+		store: s,
+	}
+
+	// Make sure we have a stable state.
+	runtime.processNewConfig()
+
+	return runtime
+}
+
+// Dispatcher returns the runtime.Dispatcher that is implemented by this runtime package.
+func (c *Runtime) Dispatcher() runtime.Dispatcher {
+	return c.dispatcher
+}
+
+// StartListening directs Runtime to start listening to configuration changes. As config changes, runtime processes
+// the confguration and creates a dispatcher.
+func (c *Runtime) StartListening() error {
+	c.stateLock.Lock()
+	defer c.stateLock.Unlock()
+
+	if c.shutdown != nil {
+		return errors.New("already listening")
```

Comment:
under what case will this happens ? Seems startListening is only going to be called once.

Reply:
Similar to the other pieces of code (i.e. server package for example), this should happen only once during the lifetime of a mixs process. However, we've been designing our other components with full life-cycle support as well (i.e. server package can be stopped/and restarted). So this is mainly for protecting against such cases.